### DISCUSSION TITLE
cannot execute: required file not found

### DISCUSSION THREAD

**Discussion by learningBASh** [termux/termux-app#3830] [Q&A]
bash: /bin/which: cannot execute: required file not found

Q.What required file could system binaries (which) need?
A. If I must guess: a lib.

**Comment by TomJo2000** [comment]
`which` is provided by the [`which` package](https://github.com/termux/termux-packages/blob/master/packages/which/build.sh).

Alternatively you may use the Bash built-in: `command -v` instead of `which` to achieve the same effect.

Also this question concerns a Termux package, not the app itself so should have been asked in [termux/termux-packages](https://github.com/termux/termux-packages).

**Comment by learningBASh** [comment]
> `which` is provided by the [`which` package](https://github.com/termux/termux-packages/blob/master/packages/which/build.sh).
> 
Usually a command not having been installed displays the pkg command to install it with the package containing the missing command. 
> Alternatively you may use the Bash built-in: `command -v` instead of `which` to achieve the same effect.
> 
command -v which appears to show which is not installed by returning nothing. Is which a special scenario in where it does not seek for the package to install and return the pkg?
> Also this question concerns a Termux package, not the app itself so should have been asked in [termux/termux-packages](https://github.com/termux/termux-packages).

Could be either. As seen this is apparently a special scenario conscerning the app as it does not return a get command for a missing command file.

**Reply by TomJo2000** [reply]
> Usually a command not having been installed displays the pkg command to install it with the package containing the missing command.

`command-not-found` works for `which` on my end.
See the section below for a more in-depth explanation.

> command -v which appears to show which is not installed by returning nothing. Is which a special scenario in where it does not seek for the package to install and return the pkg?

The suggestion for what package may contain a missing command is handled by our custom [`command-not-found`](https://github.com/termux/command-not-found) utility.
It works by calculating the [edit distance](https://en.wikipedia.org/wiki/Levenshtein_distance) of the failed command compared to a set of known commands.
I'm guessing `command -v package_name` exceeds the maximum difference for which `command-not-found` will try to make a suggestion for.

> Could be either. As seen this is apparently a special scenario conscerning the app as it does not return a get command for a missing command file.

This is not an issue with the Terminal itself, it is a nuance of the commands being run, which falls under packaging, but I mostly just wanted to point out that this isn't really the correct place to have put this question.

**Comment by learningBASh** [comment]
" I'm guessing `command -v package_name` exceeds the maximum difference for which `command-not-found` will try to make a suggestion for." If you read my report it is usual for command-not-found output to display instead of 'cannot execute: required file not found' which language may suggest the command was in place yet not running because of some dependency which is likely to be a lib. Your reply explains command-not-found is not running or giving different than usual output after executing which. As seen this is a special situation and I would take it all of the preinstalled tools would fall under termux-app and especially the custom tools like "our" being termux-app specific our command-not-found. This is not about command -v

> 
> > Could be either. As seen this is apparently a special scenario conscerning the app as it does not return a get command for a missing command file.
> 
> This is not an issue with the Terminal itself, it is a nuance of the commands being run, which falls under packaging, but I mostly just wanted to point out that this isn't really the correct place to have put this question.

Nuances of commands which are specific to the preinstalled termux-app could fall under termux-app. Whatever you are mostly trying to do you are mostly wrong. **Could be either.**

**Reply by learningBASh** [reply]
Looks like it is a malfunction here is the report https://github.com/termux/termux-packages/issues/19199#issue-2128893616

**Comment by wviana** [comment]
Hi there. I'm getting this same error trying to execute java instaled using sdkman.
I did wiped the termux application data. I'm still getting same error. Any direction? Or any thing I could do to try to figure out what is going on?

**Reply by Grimler91** [reply]
The java you installed is probably compiled against glibc, and won't work on android (well, outside of proot/chroot at least)

**Reply by Maxython** [reply]
If java is based on glibc, then it can be easily run on Termux using `grun` (`glibc-runner`).

**Reply by wviana** [reply]
Oh. Just had figured it out. By running file 
```
file java
java: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, not stripped
```

Tried to patchelf 
```
patchelf --set-interpreter /system/bin/linker64 java
```

But than I got some other errors. 

I'll try with grun as @Maxython suggested. 

Are there jvm vendors that are hardlinked and some dyn? Just wondering if some other vendor could had work.

**Reply by wviana** [reply]
Okay. I didn't find a `glibc-runner` package that contains `grun`. The closest I've found was `glibc-repo`, but it doesn't appear to have `grun` either. I'll check some other vendors later.

**Reply by Maxython** [reply]
> Oh. Just had figured it out. By running file
> 
> ```
> file java
> java: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, not stripped
> ```
> 
> Tried to patchelf
> 
> ```
> patchelf --set-interpreter /system/bin/linker64 java
> ```
> 
> But than I got some other errors.

For glibc-based binaries, you cannot configure a dynamic linker to a bionic linker. Glibc-based binaries only work with the glibc linker.

> Okay. I didn't find a glibc-runner package that contains grun. The closest I've found was glibc-repo, but it doesn't appear to have grun either. I'll check some other vendors later.

The `glibc-runner` package is in a separate dedicated repo for glibc packages. To access this repo you must install the `glibc-repo` package. Here are the commands for better understanding:
```shell
pkg ins glibc-repo -y
pkg ins glibc-runner -y
# Note: do not use your customized binary, install an unconfigured one
grun -f ./java
```

**Reply by wviana** [reply]
Thx so much for taking your time.

Now it sounds obvious that the glibc-repo I've installed was the repo and that I should had searched for the runner after that. My bad. 

So I tried grun. 
Here are my attempts and results. 
```shell
~/.../8.0.412-tem/bin $ ./java -version
bash: ./java: cannot execute: required file not found
~/.../8.0.412-tem/bin $ grun java -version
java: error while loading shared libraries: java: cannot open shared object file: No such file or directory
~/.../8.0.412-tem/bin $ grun ./java -version
Error: could not find libjava.so
Error: Could not find Java SE Runtime Environment.
~/.../8.0.412-tem/bin $ grun --help
Help message from glibc-runner v2.0

glibc-runner - launcher for working with the glibc shell or with a glibc-based binary

Options:
 --help      -h  print help message
 --info      -i  print information about the running glibc-runner shell
 --shell     -s  run the glibc-runner shell or a command from that shell
 --teg       -t  enable the use of termux-exec-glibc in the glibc-runner shell
 --configure -c  configure the binary to run on the device
 --findlib   -f  find libraries for the binary
 --no-linker -n  don't use dynamic linker to launch binary
 --debug     -d  [1|2|3|4]  launch binary or shell under strace

Example: glibc-runner [-c|-f|-n] ./binary || grun [-s|-n|-t] [gcc -v]

Bug report: https://github.com/termux-pacman/glibc-packages/issues
~/.../8.0.412-tem/bin $ grun --shell
~/.../8.0.412-tem/bin $ java -version
bash: /data/data/com.termux/files/home/.sdkman/candidates/java/current/bin/java: cannot execute: required file not found
~/.../8.0.412-tem/bin $ grun ./java -version
Error: could not find libjava.so
Error: Could not find Java SE Runtime Environment.
~/.../8.0.412-tem/bin $
```

I've tried with shell thinking it may not be setting JAVAHOME when using grun. And loading a new shell would have sdkman setting the right home. 

Right before sending this I've notice you mention using `-f` flag. 

```
~/.../8.0.412-tem/bin $ grun -f ./java -version
Message from glibc-runner: searching for libraries...
Error from glibc-runner: could not find 'libjli.so'
```

Same results. 

I've deleted all downloaded candidates and download again before trying grun. To be sure not being affected by earlier pathelf.

**Reply by Maxython** [reply]
Is there a `lib` folder in the `8.0.412-tem` directory? If so, try running `grun` from the `8.0.412-tem` directory so that `grun` can check the `lib` folder. Here is the command to run:
```shell
# don't run this with the `-version` flag
grun -f -c ./bin/java
# runs in grun shell
./bin/java -version
```

**Reply by wviana** [reply]
Awesome. 
Thx very much. It worked

```
~/.../java/8.0.412-tem $ grun -f -c ./bin/java
Message from glibc-runner: searching for libraries...
Message from glibc-runner: searching libraries was successful
~/.../java/8.0.412-tem $ ./bin/java -version
openjdk version "1.8.0_412"
OpenJDK Runtime Environment (Temurin)(build 1.8.0_412-b08)
OpenJDK 64-Bit Server VM (Temurin)(build 25.412-b08, mixed mode)
```

So grun patches it?

Could we package sdkman into termux so it will work out of the box for everybody else?

**Reply by Maxython** [reply]
> So grun patches it?

In your case, `grun` added the path to `.../8.0.412-tem/lib` in `java`, using `patchelf`, so that `java` could find all the libraries.

> Could we package sdkman into termux so it will work out of the box for everybody else?

It's hard for me to answer this question since I haven't worked with `sdkman`.
